internal/versions/rest_api_versions.go

```diff
@@ -0,0 +1,245 @@
+package versions
+
+import (
+	context "context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/openshift/assisted-service/internal/common"
+	"github.com/openshift/assisted-service/internal/oc"
+	models "github.com/openshift/assisted-service/models"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/thoas/go-funk"
+	"gorm.io/gorm"
+)
+
+type restAPIVersionsHandler struct {
+	log                      logrus.FieldLogger
+	releaseHandler           oc.Release
+	imagesLock               sync.Mutex
+	mustGatherVersions       MustGatherVersions
+	releaseImages            models.ReleaseImages
+	ignoredOpenshiftVersions []string
+	db                       *gorm.DB
+}
+
+// GetReleaseImage retrieves a release image based on a specified OpenShift version and CPU architecture.
+// If the provided OpenShift version includes a patch version, it attempts to retrieve an exact match for the release image if available.
+// For OpenShift versions specified as major.minor, it fetches the latest matching release image.
+// The function returns an error Returns an error for other formats of OpenShift version or if no matching image can be found.
+func (h *restAPIVersionsHandler) GetReleaseImage(_ context.Context, openshiftVersion, cpuArchitecture, _ string) (*models.ReleaseImage, error) {
+	// validations
+	if err := validateCPUArchitecture(cpuArchitecture); err != nil {
+		return nil, err
+	}
+	versionFormat := common.GetVersionFormat(openshiftVersion)
+	if versionFormat == common.NoneVersion || versionFormat == common.MajorVersion {
+		return nil,
+			errors.Errorf(
+				"invalid openshiftVersion '%s'. Expected format: 'major.minor' or 'major.minor.patch', optionally followed by a prerelease identifier",
+				openshiftVersion,
+			)
+	}
+
+	// If releaseImages were given to the handler in creation,
+	// it means that they should be the source for release images.
+	if h.releaseImages != nil {
+		for _, releaseImage := range h.releaseImages {
+			if *releaseImage.OpenshiftVersion == openshiftVersion &&
+				(funk.Contains(releaseImage.CPUArchitectures, cpuArchitecture) || *releaseImage.CPUArchitecture == cpuArchitecture) {
+				return releaseImage, nil
+			}
+		}
+
+		return nil, errors.Errorf("no release image found for openshiftVersion: '%s' and CPU architecture '%s'", openshiftVersion, cpuArchitecture)
```

Comment:
The thought process was as follows:
For KubeAPI mode, we want RELEASE_IMAGES to serve as cache.
For RestAPI mode, there are two scenarios:

1. In SaaS, we want to activate the release images gathering process, and use the DB as the source for release images.
2. In other scenarios (ABI for example), we don't want to activate the release images gathering, (by not supplying RELEASE_SOURCES, then we want to use RELEASE_IMAGES as the source for release images.

This is the way I implemented it. An alternative could be having the release images in the DB in rest mode anyway, and if the gathering is activated then add more releases, otherwise not. then you can only look at the DB for the source of release images in Rest mode, although I found this solution much bigger change then the current option.

Reply:
In second thought, it might be better to go with the alternative and have one source of release images for rest API, I will try it out